File: src/main/java/net/ndrei/teslacorelib/items/SpeedUpgrade.java
package net.ndrei.teslacorelib.items;

import net.minecraft.item.ItemStack;
import net.ndrei.teslacorelib.TeslaCoreLib;
import net.ndrei.teslacorelib.tileentities.ElectricMachine;
import net.ndrei.teslacorelib.tileentities.SidedTileEntity;

/**
 * Created by CF on 2017-04-08.
 */
public class SpeedUpgrade extends BaseTieredAddon {
    private int tier;

    public SpeedUpgrade(int tier) {
        super(TeslaCoreLib.MODID, TeslaCoreLib.creativeTab, "speed_tier" + tier);
        this.tier = tier;
    }

    @Override
    protected String getAddonFunction() {
        return "speed";
    }

    @Override
    protected int getTier() {
        return this.tier;
    }

    @Override
    public float getWorkEnergyMultiplier() {
        return 1.25f;
    }

    @Override
    public boolean canBeAddedTo(SidedTileEntity machine) {
        return SpeedUpgrade.canBeAddedToMachine(machine) && super.canBeAddedTo(machine);
    }

    static boolean canBeAddedToMachine(SidedTileEntity machine) {
        return ((machine instanceof ElectricMachine) && (((ElectricMachine) machine).supportsSpeedUpgrades()));
    }

    @Override
    public void onAdded(ItemStack addon, SidedTileEntity machine) {
        super.onAdded(addon, machine);

        if (machine instanceof ElectricMachine) {
            ((ElectricMachine) machine).updateWorkEnergyRate();
        }
    }

    @Override
    public void onRemoved(ItemStack addon, SidedTileEntity machine) {
        super.onRemoved(addon, machine);

        if (machine instanceof ElectricMachine) {
            ((ElectricMachine) machine).updateWorkEnergyRate();
        }
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/items/SpeedUpgrade.kt
package net.ndrei.teslacorelib.items

import net.minecraft.item.ItemStack
import net.ndrei.teslacorelib.TeslaCoreLib
import net.ndrei.teslacorelib.tileentities.ElectricMachine
import net.ndrei.teslacorelib.tileentities.SidedTileEntity

/**
 * Created by CF on 2017-06-27.
 */
abstract class SpeedUpgrade(override val tier: Int) : BaseTieredAddon(TeslaCoreLib.MODID, TeslaCoreLib.creativeTab, "speed_tier" + tier) {
    override val addonFunction: String
        get() = "speed"

    override val workEnergyMultiplier: Float
        get() = 1.25f

    override fun canBeAddedTo(machine: SidedTileEntity): Boolean {
        return SpeedUpgrade.canBeAddedToMachine(machine) && super.canBeAddedTo(machine)
    }

    override fun onAdded(addon: ItemStack, machine: SidedTileEntity) {
        super.onAdded(addon, machine)
        (machine as? ElectricMachine)?.updateWorkEnergyRate()
    }

    override fun onRemoved(addon: ItemStack, machine: SidedTileEntity) {
        super.onRemoved(addon, machine)
        (machine as? ElectricMachine)?.updateWorkEnergyRate()
    }

    companion object {
        internal fun canBeAddedToMachine(machine: SidedTileEntity): Boolean {
            return machine is ElectricMachine && machine.supportsSpeedUpgrades()
        }
    }
}